[Feature] Allow multiple "Insertion in the form" (dom) blocks per entity

JeremieMercier opened the pull request:
This PR provides the ability to define multiple "Insertion in the form" blocks for the same item based on its entity

- Created a new method findContainers() (based on findContainer()) that returns all 'dom' containers for an item based on its entity (with parent entity handling via getAncestorsOf()).
- Adapted hooks (pre_item_add, pre_item_update, post_item_add, post_item_update) to manage multiple containers using the _plugin_fields_data_multi array.
- Updated the populateData() function to extract input values by stripping the prefix, ensuring that data is saved into the correct columns of the injection table.
- Modified the container.form.php file to "clean" the form data (by removing the prefix) before calling updateFieldsValues(), thereby enabling the saving of domtab containers.

Please test these modifications and verify that everything works as expected for all object types supported by the Fields plugin. Note that my tests were successful, although I focused exclusively on 'Ticket' type objects for my use case.

## Checklist before requesting a review

*Please delete options that are not relevant.*

- [x] I have performed a self-review of my code.
- [ ] I have added tests (when available) that prove my fix is effective or that my feature works.
- [ ] I have updated the CHANGELOG with a short functional description of the fix or new feature.
- [x] This change requires a documentation update.

Issues : #744, #789, #834

stonebuzz commented:
Hi @JeremieMercier 

I am questioning the plugin's ability to reconcile the correct container during API calls, an issue we have encountered in the past. For example, when updating a ticket and including the "fields" in the payload, could you verify this point?

JeremieMercier commented:
I retrieve the block row ID like this:
![image](https://github.com/user-attachments/assets/653807c5-b303-41c6-ac4f-b2d084fb1f88)

Then, I update the field using the following method:
![image](https://github.com/user-attachments/assets/beb7c182-7d0e-401d-9055-3f4c22ea9f3f)

So I haven’t encountered any issue with the API — unless I misunderstood how you're updating the fields via the API?

stonebuzz commented:
It is also possible to update fields from the fields plugin during ticket update.

`PUT {{URL_GLPI}} /Ticket/5`

JeremieMercier commented:
Yes, it works!

Just include the block ID in the field name, like:
plugin_fields_15_bloconechamponefield

![image](https://github.com/user-attachments/assets/91dd1dda-38b5-4449-acf3-9ce41099406c)
![image](https://github.com/user-attachments/assets/7feb5eb9-1cff-41d3-9776-397ad6ece635)

JeremieMercier commented:
Hello, do you have time to review the PR ?

trasher commented on inc/container.class.php:
```diff
@@ -1604,6 +1588,83 @@ public static function findContainer($itemtype, $type = 'tab', $subtype = '')
+                }
+                $entityIdList = implode(",", $entityIds);
+
+                $sql = "SELECT id FROM glpi_plugin_fields_containers
```
You must use [GLPI DBIterator](https://glpi-developer-documentation.readthedocs.io/en/develop/devapi/database/dbiterator.html) instead of direct queries.

JeremieMercier commented:
A syntax error, it was slipped in after applying the suggestion 😅

JeremieMercier commented:
Are we good this time ? 😅🤞

JeremieMercier commented:
Hi, what's the next step ?

stonebuzz commented:
Hi @JeremieMercier 

The feature you are proposing is substantial and introduces significant changes. To ensure its smooth integration and functionality, we must be extremely vigilant. This means we need to "lock down" our testing to guarantee that everything works as expected.

It is important to acknowledge that the plugin is already complex to maintain, especially with the new features introduced in GLPI. Given this complexity, we must approach any additional features with caution to ensure they do not further complicate maintenance and support.

To ensure the proper functioning of this PR, here is a concise list of unit tests that should be added:

### Unit Tests for `findContainers`

1. **Retrieve Containers for Specific Itemtype and Entity**
   - Verify that `findContainers` returns the correct containers for a given itemtype and entity.

2. **Retrieve Containers with Subtype**
   - Verify that `findContainers` returns the correct containers when a subtype is specified.

3. **Retrieve Containers with Recursive Entity**
   - Verify that `findContainers` considers recursive entities.

### Unit Tests for Hooks

4. **`preItemAdd` with Multiple Containers**
   - Verify that the `preItemAdd` hook correctly handles data for multiple containers.

5. **`postItemAdd` with Multiple Containers**
   - Verify that the `postItemAdd` hook correctly updates field values for multiple containers.

6. **`preItemUpdate` with Multiple Containers**
   - Verify that the `preItemUpdate` hook correctly handles data for multiple containers during item update.

### Unit Tests for `populateData`

7. **`populateData` with Prefix**
   - Verify that `populateData` correctly extracts field values by removing the prefix.

8. **`populateData` with Multiple Selection Fields**
   - Verify that `populateData` correctly handles fields allowing multiple selections.

### Unit Tests for `showForTab`

9. **`showForTab` with Multiple Containers**
   - Verify that `showForTab` correctly displays containers for a given item.

10. **`showForTab` with Insufficient Rights**
    - Verify that `showForTab` returns nothing if the user does not have sufficient rights on the containers.

### Unit Tests for `API`

11. **Ticket input update**

12. **Container update**

JeremieMercier commented:
Hi @stonebuzz Thanks for the message

I fully understand the concerns and the importance of thorough testing.
I'm currently working on implementing the required tests and will make sure all the listed cases are properly covered.

While working on the tests already partially in place, I’ve been able to identify and fix some issues especially related to "dropdown" field handling and container restriction management.

I’ll keep updating here as I progress.

akstis-typer commented:
Hi, @JeremieMercier.
What tool you are using to send HTTP requests?
I'm newby in WEB development so I'm not entirely shure what tool is that.
Thank you :)

JeremieMercier commented:
I’ve added all the requested unit tests (except the API ones):

- findContainers : basic cases, sub-types, entity recursion
- Hooks: preItem, postItemAdd, preItemUpdate with multiple containers
- populateData: prefix handling and multi-selection fields
- showForTab: displaying multiple containers and checking rights

Please note I’m not a testing expert, so the tests may not follow every usual best practice. Feel free to share any feedback so I can improve.

The API tests still need to be written by whoever is willing to tackle them.

CorentinS6 commented:
And code generated by AI ?

JeremieMercier commented:
> And code generated by AI ?

Hello, All the logic is 100% mine, however for the tests, it's another story, I used AI to learn and understand the use of the tests in addition to videos like this one from Grafikart, but as I said, I'm not an expert..., the tests can also be redone by a more competent person if necessary knowing that basically, there were no tests set up on the plugin.

CorentinS6 commented:
@stonebuzz , is this work good enough for the merge now?

Rom1-B commented on front/container.form.php:
```diff
@@ -57,6 +57,17 @@
+                $data[$key] = $value;
+            }
+        }
         $container->updateFieldsValues($_REQUEST, $_REQUEST['itemtype'], false);
```
`$data` is computed but never used?!

JeremieMercier replied:
yes that was a true bug : the prefix-stripping loop built $data but the call still passed $_REQUEST, so prefixed keys were never matched by updateFieldsValues() (which expects unprefixed keys). fixed the cleaned $data is now passed.

Rom1-B commented on inc/container.class.php:
```diff
@@ -1774,68 +1845,57 @@ public static function preItemUpdate(CommonDBTM $item)
-        if ($item->isEntityAssign() && !in_array($item->getEntityID(), $entities)) {
-            return false;
-        }
+            // need to check if container is usable on this object entity
+            $entities = [$loc_c->fields['entities_id']];
+            if ($loc_c->fields['is_recursive']) {
+                $entities = getSonsOf(getTableForItemType('Entity'), $loc_c->fields['entities_id']);
+            }
 
-        if (false !== ($data = self::populateData($c_id, $item))) {
-            if (self::validateValues($data, $item::getType(), isset($_REQUEST['massiveaction'])) === false) {
-                $item->input = [];
+            if (count($item->fields) === 0) {
+                $item->fields = $item->input;
+            }
 
-                return false;
+            if ($item->isEntityAssign() && !in_array($item->getEntityID(), $entities)) {
+                continue; // not the right entity
             }
 
-            $item->input['_plugin_fields_data'] = $data;
+            if (false !== ($data = self::populateData($c_id, $item))) {
+                if (!self::validateValues($data, $item->getType(), isset($_REQUEST['massiveaction']))) {
+                    // if validation fails, we need to remove the data from the item input
+                    $item->input = [];
 
-            return true;
+                    return false;
+                }
+                $all_data[] = $data;
+            }
```
When an item has several applicable containers, a validation failure on any single container wipes `$item->input` and returns immediately, discarding the data already accumulated in `$all_data` for the other, valid containers. In the single-container world this was the existing behaviour (nothing else to lose), but with multiple containers per item it means one invalid field in container B now also blocks saving valid data in container A. Please confirm whether this all-or-nothing semantics is intentional; if not, previously-validated containers' data should still make it into `_plugin_fields_data_multi`.

JeremieMercier replied:
Yes the all-or-nothing semantics is intentional all applicable blocks are rendered and submitted within a single form, so a validation failure anywhere blocks the whole submission the user gets the error messages from validateValues(), fixes the input and resubmits. Partially saving would be worse : the item would be created with container A's data while container B's data is silently dropped. This also matches the historical single-container behaviour, simply extended to several containers

Rom1-B commented on inc/field.class.php:
```diff
@@ -863,8 +863,13 @@ public static function showDomContainer($id, $item, $type = 'dom', $subtype = ''
         echo Html::hidden('_plugin_fields_type', ['value' => $type]);
         echo Html::hidden('_plugin_fields_subtype', ['value' => $subtype]);
+        echo Html::hidden($fieldTypeName, ['value' => $type]);
+        echo Html::hidden($fieldSubTypeName, ['value' => $subtype]);
```
`_plugin_fields_type_{id}` / `_plugin_fields_subtype_{id}` are added alongside the existing non-suffixed hidden fields, but no other code (PHP, Twig, or JS) in the plugin reads them back. Either wire them up (presumably needed to disambiguate type/subtype per container once several `dom` containers coexist on one page) or drop them.

JeremieMercier replied:
You're right they were dead code an early attempt at per-container disambiguation that became unnecessary once preItem() resolved applicable containers itself via findContainers(). They have been dropped only the non-suffixed _plugin_fields_type / _plugin_fields_subtype fields remain

Rom1-B commented on composer.lock:
The diff downgrades `sebastian/diff` from `6.0.2` (PHP `>=8.2`, dev-requires `phpunit/phpunit ^11.0`) to `4.0.6` (PHP `>=7.3`, dev-requires `phpunit/phpunit ^9.3`), alongside an unrelated `phpstan/phpstan` patch bump. This looks like stale lockfile drift from an out-of-sync branch rather than an intentional change for this feature. Suggest regenerating the lockfile with a clean `composer update` against current `composer.json` before merge.

Rom1-B commented on tests/PluginFieldsContainerTest.php:
```diff
@@ -0,0 +1,753 @@
+            'entities_id'                               => 0,
+        ]);
+
+        // 3. Injecter les données comme si elles venaient du formulaire
```
Should be in English per the project's coding conventions.

JeremieMercier committed:
[Feature] Allow multiple "Insertion in the form" (dom) blocks per entity

Provide the ability to define multiple 'dom' blocks for the same item
based on its entity.

- New PluginFieldsContainer::findContainers() returning all active 'dom'
  containers applicable to an item and its entity (with parent entity
  handling via getAncestorsOf(), recursivity and profile rights checks).
- Hooks (pre_item_add, pre_item_update, post_item_add) handle multiple
  containers through the _plugin_fields_data_multi array.
- populateData() extracts input values from per-container prefixed keys
  (plugin_fields_{container_id}_...), with fallback to unprefixed keys.
- fields.html.twig prefixes input names with the container id.
- container.form.php strips the container prefix before calling
  updateFieldsValues(), enabling domtab containers saving.
- showForTab() renders every applicable container, each with its own
  scoped refresh script.
- Add PHPUnit coverage: findContainers (subtypes, entity recursion),
  multi-container hooks, populateData prefix and multi-selection
  handling, showForTab display and rights.

Closes #744, #789, #834

Rom1-B commented on inc/container.class.php:
```diff
@@ -1891,87 +1959,76 @@ public static function preItemUpdate(CommonDBTM $item)
-        if (false !== ($data = self::populateData($c_id, $item))) {
-            if (self::validateValues($data, $item::getType(), isset($_REQUEST['massiveaction'])) === false) {
-                $item->input = [];
+            if (false !== ($data = self::populateData($c_id, $item))) {
+                if (!self::validateValues($data, $item->getType(), isset($_REQUEST['massiveaction']))) {
+                    // if validation fails, we need to remove the data from the item input
+                    $item->input = [];
 
-                return false;
+                    return false;
+                }
+                $all_data[] = $data;
+                continue;
             }
```
`preItem()` loops over every `dom` container applicable to the item. Before this PR there was only ever one container per itemtype, so returning `false` on the first validation failure was harmless. Now that several containers can apply to the same item, if container A is valid and container B is not, the loop still wipes `$item->input` and returns `false` as soon as B fails ; discarding A's already-validated data too. One invalid container therefore blocks saving every other valid container on the same item.

@JeremieMercier is this all-or-nothing behavior intentional, or should valid containers still be saved when another one fails validation? If the latter, previously-validated containers' data should still be pushed into `_plugin_fields_data_multi` instead of being wiped.

JeremieMercier replied:
Yes the all-or-nothing semantics is intentional all applicable blocks are rendered and submitted within a single form, so a validation failure anywhere blocks the whole submission the user gets the error messages from validateValues(), fixes the input and resubmits. Partially saving would be worse : the item would be created with container A's data while container B's data is silently dropped. This also matches the historical single-container behaviour, simply extended to several containers

Rom1-B reviewed (changes requested):
Several errors in the CI

JeremieMercier committed:
[CI] Add CHANGELOG entry, apply php-cs-fixer and rector fixes

JeremieMercier committed:
[Fix] Restore legacy field input handling and fix multi-dom regressions

- populateData(): accept bare field names (API/external callers) as fallback
  of the new prefixed input names
- populateData(): treat null input values as not provided (isset semantics),
  fixes SQL error on empty dropdown submitted from a form destination
- preItem(): restore explicit c_id support (tab forms, form destinations)
- findContainers(): JSON-encode itemtype for JSON_CONTAINS to support
  namespaced itemtypes on strict MySQL
- richtext companion inputs stored under the column-based key expected
  by addRichTextFiles()
- tests: multiple dom containers per itemtype are now expected to be allowed
- tests: do not leak $_REQUEST['massiveaction'] between test files

JeremieMercier committed:
Merge remote-tracking branch 'upstream/main' into feature/multiple-dom-blocks

# Conflicts:
#	CHANGELOG.md

JeremieMercier commented:
hi @Rom1-B, and now ?
